File: solutions/dedup-redup.py
from collections import OrderedDict
import os

# Complete the functions below.
CHUNK_SIZE = 1024
INT_SIZE = 32   
ENCODING = 'cp1252'


def dedup(input_file_path, output_file_path):
    # Open the files
    try:
        in_file = open(input_file_path, 'rb')
        out_file = open(output_file_path, 'wb')
    except Exception:
        print('File open failed!')
    
    # Find the file sizes
    in_file_size = os.path.getsize(input_file_path)
    n_chunks = int(in_file_size / CHUNK_SIZE)
    
    # We maintain:
    #  - unique_chunk_map: Map of chunks to indices.
    #  - chunk_index_list: List of indices in the unique_chunk map.
    unique_chunk_map = OrderedDict()
    chunk_index_list = [None] * n_chunks
    
    # read the file chunk by chunk
    for ix in range(n_chunks):
        chunk = in_file.read(CHUNK_SIZE)

        # If not found earlier, insert in the map.
        if chunk not in unique_chunk_map:
            unique_chunk_map[chunk] = len(unique_chunk_map)
        
        # Add the index of the unique chunk found
        chunk_index_list[ix] = unique_chunk_map[chunk]

    # Write the number of chunks to the BEGINNING of the file
    out_file.write(len(unique_chunk_map).to_bytes(INT_SIZE, byteorder='big'))
    
    for chunk in unique_chunk_map.keys():
        out_file.write(chunk)
        
    out_file.write(','.join([str(x) for x in chunk_index_list]).encode(ENCODING))

    # Close the files
    in_file.close()
    out_file.close()
    

def redup(input_file_path, output_file_path):
    # Open the files
    try:
        in_file = open(input_file_path, 'rb')
        out_file = open(output_file_path, 'wb')
    except Exception:
        print('File open failed!')
    
    # We maintain:
    #  - unique_chunk_map: Map of chunks to indices.
    #  - chunk_index_list: List of indices in the unique_chunk map.
    unique_chunk_map_inv = OrderedDict()
    
    # Get the number of unique chunks from the first INT_SIZE bytes of the file
    n_unique_chunks = int.from_bytes(in_file.read(INT_SIZE), byteorder='big')
    
    # Get all the unique chunks
    for ix in range(n_unique_chunks):
        unique_chunk_map_inv[ix] = in_file.read(CHUNK_SIZE)

    # Get the list: the rest of the bytes would just contain that!
    list_as_str = in_file.read().decode(ENCODING)
    chunk_index_list = [int(x) for x in list_as_str.split(',')]
    
    
    # Write the chunks!
    for chunk_index in chunk_index_list:
        out_file.write(unique_chunk_map_inv[chunk_index])

    # Close the files
    in_file.close()
    out_file.close()
    
    with open(output_file_path, 'rb') as infile:
        output_contents = infile.read()
    
    return True

File: solutions/dedup-redup.cpp
#include <fstream>

int CHUNK_SIZE = 1024;

void dedup(std::string inpath, std::string outpath)
{
    char chunk[CHUNK_SIZE];
    
    // Open input file
    std::ifstream infile(inpath);

    // Represent input file as a hash table:
    //   - key: hash of a unique 1KB chunk
    //   - value: list of indices within the file where the chunk is found
    int64_t hash;
    auto table = std::map<int64_t, std::vector<int>>();
    for (int index = 0; !infile.eof(); index++) {
        infile.read(chunk, CHUNK_SIZE);
        int count = infile.gcount();
        if (!count)
            break;
        
        // compute hash
        std::string str(chunk, count);        
        hash = std::hash<std::string>{}(str);
        
        // check if hash already exists
        auto it = table.find(hash);
        if (it == table.end()) {
            table[hash] = std::vector<int>{index};
            continue;
        }
        
        // append to list of indices
        it->second.push_back(index);
    }

    // Open output files
    std::ofstream outfile1(outpath);
    std::ofstream outfile2(outpath);

    // Write out hash table into the output file.
    // Here's how it's organized:
    //   - first four bytes: number of chunks
    //   - next 1024 bytes: first chunk
    //   - next 1024 bytes: second chunk
    //   - ...
    //   - comma-separated list of indices for first chunk (ending in newline)
    //   - comma-separated list of indices for second chunk (ending in newline)
    //   - ...
    //   - checksum
    int32_t chunk_count = table.size();
    outfile1.write((char *) &chunk_count, sizeof(int32_t));
    outfile2.seekp(sizeof(int32_t) + (chunk_count * CHUNK_SIZE));
    
    infile.clear();
    infile.seekg(0);
    while (!infile.eof()) {
        infile.read(chunk, CHUNK_SIZE);
        int count = infile.gcount();
        if (!count)
            break;
        
        // compute hash
        std::string str(chunk, count);
        int64_t hash = std::hash<std::string>{}(str);
        
        // check if chunk is still in hash table
        auto it = table.find(hash);
        if (it == table.end())
            continue;

        // get indices
        std::string indices;
        for (const auto &index : it->second) {
            if (&index != &it->second[0])
                indices += ",";
            indices += std::to_string(index);
        }
        indices += "\n";        
        
        // write out chunk and indices
        outfile1.write(chunk, CHUNK_SIZE);
        outfile2.write(indices.c_str(), indices.length());
        
        // remove chunk from hash table since it's been processed
        table.erase(it);
    }
    
    // Close files
    infile.close();
    outfile1.close();
    outfile2.close();
    
    // Calculate checksum of the output file
    std::fstream outfile(outpath, std::ios_base::in | std::ios_base::out);
    hash = 0;
    while (!outfile.eof()) {
        outfile.read(chunk, CHUNK_SIZE);
        int count = outfile.gcount();
        if (!count)
            break;

        // compute hash
        std::string str(chunk, outfile.gcount());        
        hash ^= std::hash<std::string>{}(str);
    }

    outfile.clear();
    outfile.write((char *) &hash, sizeof(int64_t));
    outfile.close();
}

bool redup(std::string inpath, std::string outpath)
{
    char chunk[CHUNK_SIZE];

    // Open input file
    std::ifstream infile(inpath, std::ifstream::ate);
    int file_size = infile.tellg();
    infile.clear();
    infile.seekg(0);

    // Compute sizes of the data vs. checksum in the input file
    int checksum_bytes = sizeof(int64_t);
    int data_bytes = file_size - checksum_bytes;

    // Verify the integrity of the input file by recomputing the checksum
    int64_t hash = 0;
    while (data_bytes > 0) {
        infile.read(chunk, std::min(CHUNK_SIZE, data_bytes));
        int count = infile.gcount();
        if (!count)
            break;
        
        std::string str(chunk, count);
        hash ^= std::hash<std::string>{}(str);
        
        data_bytes -= count;
    }
    
    // Compare checksums: recomputed vs. original
    infile.seekg(-checksum_bytes, infile.end);
    int64_t checksum;
    infile.read((char *) &checksum, sizeof(int64_t));
    infile.close();    
    if (hash != checksum)
        return false;

    int32_t chunk_count;
    
    // Open input and output files
    std::ifstream infile1(inpath);
    infile1.read((char *) &chunk_count, sizeof(int32_t));
    
    std::ifstream infile2(inpath);
    infile2.seekg(sizeof(int32_t) + chunk_count * CHUNK_SIZE);    
    
    std::ofstream outfile(outpath);
    
    // Write out the individual chunks to the output file
    std::string line;    
    for (int i = 0; i < chunk_count; i++) {
        infile1.read(chunk, CHUNK_SIZE);
        std::getline(infile2, line);
        
        std::stringstream stream(line);
        std::string index;
        while (std::getline(stream, index, ',')) {
            outfile.seekp(std::stoi(index) * CHUNK_SIZE);
            outfile.write(chunk, CHUNK_SIZE);            
        }
    }
    
    // Close input and output files
    infile1.close();
    infile2.close();
    outfile.close();
    
    return true;
}
